README.md
# NAME

Web::ChromeLogger - ChromeLogger for Perl

# SYNOPSIS

    use Web::ChromeLogger;

    get '/', sub {
        my $logger = Web::ChromeLogger->new();
        $logger->info('hey!');
        $logger->infof('Hello, %s!', 'John');

        my $html = render_html();

        return [
            200,
            ['X-ChromeLogger-Data' => $logger->finalize()],
            $html,
        ];
    };

# DESCRIPTION

Web::ChromeLogger is a ChromeLogger library for Perl5.

Chrome Logger is a Google Chrome extension for debugging server side applications in the Chrome console.

# HOW IT WORKS

This module generates the string for ChromeLogger. You can send the string in 'X-ChromeLogger-Data' header in HTTP response.

For more details, please see [ChromeLogger's Technical Specification](http://craig.is/writing/chrome-logger/techspecs)

# LARGE RESPONSE HEADER

## NGINX PROXY

If you are using nginx as reverse proxy, you may need to set following parameters in your configuration file:

    proxy_buffer_size   128k;
    proxy_buffers   4 256k;
    proxy_busy_buffers_size   256k;

## Maximum Data Size

Chrome has a limit of `250kb` across all headers for a single request so that is the maximum amount of encoded data you can send.

# METHODS

- `my $logger = Web::ChromeLogger->new(%args)`

    Create new instance with following parameters:

    - json\_encoder (Default: `JSON::XS->new()->ascii(1)->convert_blessed` )

        JSON encoder object. You can use JSON::XS or JSON::PP for this.

        I guess you don't need to set this parameter.

- `$logger->group($title: Str)`
- `$logger->groupf($format: Str, @list)`

    Push `group`.

- `$logger->group_end($title: Str)`
- `$logger->group_endf($format: Str, @list)`

    Push `groupEnd`.

- `$logger->group_collapsed($title: Str)`
- `$logger->group_collapsedf($format: Str, @list)`

    Push `groupCollapsed`.

- `$logger->info($title: Str)`
- `$logger->infof($format: Str, @list)`

    Push `info`.

- `$logger->warn($title: Str)`
- `$logger->warnf($format: Str, @list)`

    Push `warn`.

- `$logger->error($title: Str)`
- `$logger->errorf($format: Str, @list)`

    Push `error`.

- `$logger->wrap_by_group($title: Str)`
- `$logger->wrap_by_groupf($format: Str, @list)`

    Wrap current logging data by `$title` group.

- `$logger->finalize()`

    Generate header string.

# LICENSE

Copyright (C) Tokuhiro Matsuno.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Tokuhiro Matsuno <tokuhirom@gmail.com>

moznion

# SEE ALSO

[ChromeLogger](http://craig.is/writing/chrome-logger)


Changes
Revision history for Perl extension Web-ChromeLogger

{{$NEXT}}

0.06 2016-02-28T03:33:47Z

    commit 4f385f36aaea9ef917673835deda9a354f174aad
    Author: Kang-min Liu <gugod@gugod.org>
    Date:   Thu Feb 25 12:58:11 2016 +0100

        Fix encode() method mis-implementation.

        The ChromeLogger tech spec:

            https://craig.is/writing/chrome-logger/techspecs

        ... specified the base64 encoded result must contain no
        newline characters:

            final_data = utf8_data.encode('base64').replace('\n', '')

        The newline should be *removed* instead of being replaced by a pair of
        single quotes.
0.05 2014-04-07T04:56:25Z

    - Added Web::ChromeLogger::Null
      (moznion)
    - Set 'alllow_nonref' option for JSON encoder.
      (tokuhirom)
    - Implement formatted methods
      (moznion)

0.04 2014-03-30T10:13:25Z

    commit 6b0776ae8d67e93fc87b797e61521597eb29f7cf
    Author: bayashi <bayashi@cpan.org>
    Date:   Sun Mar 30 12:56:23 2014 +0900

    support JSON::XS 2.X
    JSON::XS 2.X exports the 'to_json' method.
    It's also implemeted in Web::ChromeLogger.

0.03 2014-03-28T01:10:23Z

    - Fixed dependencies.

0.02 2014-03-28T00:42:18Z

    - Better POD markup.

0.01 2014-03-28T00:38:24Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Web-ChromeLogger',
    module_name     => 'Web::ChromeLogger',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


author/generate_nop.pl
#!/usr/bin/env perl

# This script generates `/lib/Web/ChromeLogger/NOP.pm`.

use strict;
use warnings;
use Class::Inspector;
use FindBin;
use lib "$FindBin::Bin/../lib";
use Web::ChromeLogger;

my @public_methods = grep { $_ !~ /(?:new|finalize)/ } @{Class::Inspector->methods('Web::ChromeLogger', 'public')};

my $nop_source = <<'...';
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY `author/generate_nop.pl`.
# DO NOT EDIT DIRECTLY.
# =========================================================================

package Web::ChromeLogger::NOP;
use strict;
use warnings;
use Carp ();

sub new {
    my $class = shift;
    bless {}, $class;
}

sub finalize {
    Carp::croak "Web::ChromeLogger::NOP cannot finalize logs";
}

...

for my $public_method (@public_methods) {
    $nop_source .= <<"...";
sub $public_method {}

...
}

$nop_source .= <<'...';
1;
__END__

=encoding utf-8

=head1 NAME

Web::ChromeLogger::NOP - Dummy of Web::ChromeLogger

=head1 SYNOPSIS

    use Web::ChromeLogger::NOP;

    get '/', sub {
        my $logger = Web::ChromeLogger::NOP->new();
        $logger->info('hey!'); # NOP!

        my $html = render_html();

        return [
            200,
            ['X-ChromeLogger-Data' => $logger->finalize()], # NOP!
            $html,
        ];
    };

=head1 DESCRIPTION

Web::ChromeLogger::NOP is a dummy of C<Web::ChromeLogger>.

This class provides methods that don't work anything.

=head1 METHODS

=over 4

=item C<< my $logger = Web::ChromeLogger::NOP->new() >>

Returns instance of Web::ChromeLogger::NOP.

=item C<< $logger->finalize() >>

Always throws exception.

=back

And other provided methods are the same as C<Web::ChromeLogger>, but they don't work anything.

=head1 SEE ALSO

C<Web::ChromeLogger>

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

...

open my $fh, '>', "$FindBin::Bin/../lib/Web/ChromeLogger/NOP.pm";
print $fh $nop_source;



author/generate_null.pl
#!/usr/bin/env perl

# This script generates `/lib/Web/ChromeLogger/Null.pm`.

use strict;
use warnings;
use Class::Inspector;
use FindBin;
use lib "$FindBin::Bin/../lib";
use Web::ChromeLogger;

my @public_methods = grep { $_ !~ /(?:new|finalize)/ } @{Class::Inspector->methods('Web::ChromeLogger', 'public')};

my $source = <<'...';
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY `author/generate_null.pl`.
# DO NOT EDIT DIRECTLY.
# =========================================================================

package Web::ChromeLogger::Null;
use strict;
use warnings;
use Carp ();

sub new {
    my $class = shift;
    bless {}, $class;
}

sub finalize {
    Carp::croak "Web::ChromeLogger::Null cannot finalize logs";
}

...

for my $public_method (@public_methods) {
    $source .= <<"...";
sub $public_method {}

...
}

$source .= <<'...';
1;
__END__

=encoding utf-8

=head1 NAME

Web::ChromeLogger::Null - Dummy of Web::ChromeLogger

=head1 SYNOPSIS

    use Web::ChromeLogger::Null;

    get '/', sub {
        my $logger = Web::ChromeLogger::Null->new();
        $logger->info('hey!'); # NOP

        my $html = render_html();

        return [
            200,
            ['X-ChromeLogger-Data' => $logger->finalize()], # Throw exception
            $html,
        ];
    };

=head1 DESCRIPTION

Web::ChromeLogger::Null is a dummy of C<Web::ChromeLogger>.

This class provides methods that don't work anything.

=head1 METHODS

=over 4

=item C<< my $logger = Web::ChromeLogger::Null->new() >>

Returns instance of Web::ChromeLogger::Null.

=item C<< $logger->finalize() >>

Always throws exception.

=back

And other provided methods are the same as C<Web::ChromeLogger>, but they don't work anything.

=head1 SEE ALSO

C<Web::ChromeLogger>

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

...

open my $fh, '>', "$FindBin::Bin/../lib/Web/ChromeLogger/Null.pm";
print $fh $source;



lib/Web/ChromeLogger/NOP.pm
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY `author/generate_nop.pl`.
# DO NOT EDIT DIRECTLY.
# =========================================================================

package Web::ChromeLogger::NOP;
use strict;
use warnings;
use Carp ();

sub new {
    my $class = shift;
    bless {}, $class;
}

sub finalize {
    Carp::croak "Web::ChromeLogger::NOP cannot finalize logs";
}

sub encode {}

sub error {}

sub group {}

sub group_collapsed {}

sub group_end {}

sub info {}

sub json_encoder {}

sub push_log {}

sub to_json {}

sub unshift_log {}

sub warn {}

sub wrap_by_group {}

1;
__END__

=encoding utf-8

=head1 NAME

Web::ChromeLogger::NOP - Dummy of Web::ChromeLogger

=head1 SYNOPSIS

    use Web::ChromeLogger::NOP;

    get '/', sub {
        my $logger = Web::ChromeLogger::NOP->new();
        $logger->info('hey!'); # NOP!

        my $html = render_html();

        return [
            200,
            ['X-ChromeLogger-Data' => $logger->finalize()], # NOP!
            $html,
        ];
    };

=head1 DESCRIPTION

Web::ChromeLogger::NOP is a dummy of C<Web::ChromeLogger>.

This class provides methods that don't work anything.

=head1 METHODS

=over 4

=item C<< my $logger = Web::ChromeLogger::NOP->new() >>

Returns instance of Web::ChromeLogger::NOP.

=item C<< $logger->finalize() >>

Always throws exception.

=back

And other provided methods are the same as C<Web::ChromeLogger>, but they don't work anything.

=head1 SEE ALSO

C<Web::ChromeLogger>

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/Web/ChromeLogger/Null.pm
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY `author/generate_null.pl`.
# DO NOT EDIT DIRECTLY.
# =========================================================================

package Web::ChromeLogger::Null;
use strict;
use warnings;
use Carp ();

sub new {
    my $class = shift;
    bless {}, $class;
}

sub finalize {
    Carp::croak "Web::ChromeLogger::Null cannot finalize logs";
}

sub encode {}

sub error {}

sub group {}

sub group_collapsed {}

sub group_end {}

sub info {}

sub json_encoder {}

sub push_log {}

sub to_json {}

sub unshift_log {}

sub warn {}

sub wrap_by_group {}

1;
__END__

=encoding utf-8

=head1 NAME

Web::ChromeLogger::Null - Dummy of Web::ChromeLogger

=head1 SYNOPSIS

    use Web::ChromeLogger::Null;

    get '/', sub {
        my $logger = Web::ChromeLogger::Null->new();
        $logger->info('hey!'); # NOP

        my $html = render_html();

        return [
            200,
            ['X-ChromeLogger-Data' => $logger->finalize()], # Throw exception
            $html,
        ];
    };

=head1 DESCRIPTION

Web::ChromeLogger::Null is a dummy of C<Web::ChromeLogger>.

This class provides methods that don't work anything.

=head1 METHODS

=over 4

=item C<< my $logger = Web::ChromeLogger::Null->new() >>

Returns instance of Web::ChromeLogger::Null.

=item C<< $logger->finalize() >>

Always throws exception.

=back

And other provided methods are the same as C<Web::ChromeLogger>, but they don't work anything.

=head1 SEE ALSO

C<Web::ChromeLogger>

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/Web/ChromeLogger.pm
package Web::ChromeLogger;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.05";

use JSON::XS qw//;
use MIME::Base64 qw//;

sub new {
    my $class = shift;
    my %args = @_==1 ? %{$_[0]} : @_;
    my $self  = bless {
        %args
    }, $class;
    # allow_unknown: Should not throw an exception if when it encounters values it cannot represent in JSON.
    $self->{'json_encoder'} ||= JSON::XS->new()->ascii(1)->convert_blessed->allow_unknown;
    $self->{'logs'} = [];
    return $self;
}

sub json_encoder { $_[0]->{json_encoder} }

sub group {
    my $self = shift;
    $self->push_log('group', @_);
}

sub groupf {
    my $self   = shift;
    my $format = shift;
    $self->group(sprintf $format, @_);
}

sub group_end {
    my $self = shift;
    $self->push_log('groupEnd', @_);
}

sub group_endf {
    my $self   = shift;
    my $format = shift;
    $self->group_end(sprintf $format, @_);
}

sub group_collapsed {
    my $self = shift;
    $self->push_log('groupCollapsed', @_);
}

sub group_collapsedf {
    my $self   = shift;
    my $format = shift;
    $self->group_collapsed(sprintf $format, @_);
}

sub info {
    my $self = shift;
    $self->push_log('info', @_);
}

sub infof {
    my $self   = shift;
    my $format = shift;
    $self->info(sprintf $format, @_);
}

sub warn {
    my $self = shift;
    $self->push_log('warn', @_);
}

sub warnf {
    my $self   = shift;
    my $format = shift;
    $self->warn(sprintf $format, @_);
}

sub error {
    my $self = shift;
    $self->push_log('error', @_);
}

sub errorf {
    my $self   = shift;
    my $format = shift;
    $self->error(sprintf $format, @_);
}

# User can overwrite this method in child class.
sub to_json {
    my ($self, $stuff) = @_;
    "$stuff"
}

sub encode {
    my ($self, $rows) = @_;

    no warnings 'once';
    local *UNIVERSAL::TO_JSON = sub { $self->to_json(@_) };
    my $json_data = $self->json_encoder->encode(
        {
            "version" => "0.2",
            "columns" => [ "log", "backtrace", "type" ],
            "rows"    => $rows,
        },
    );
    my $mime_data = MIME::Base64::encode_base64($json_data);
    $mime_data =~ s/\n//g;

    return $mime_data;
}

sub wrap_by_group {
    my ($self, $title) = @_;
    $self->unshift_log('group', $title);
    $self->push_log('groupEnd', $title);
}

sub wrap_by_groupf {
    my $self   = shift;
    my $format = shift;
    $self->wrap_by_group(sprintf $format, @_);
}

sub push_log {
    my $self = shift;
    push @{ $self->{logs} }, [ [ $_[1] ], $_[2], $_[0] ];
}

sub unshift_log {
    my $self = shift;
    unshift @{ $self->{logs} }, [ [ $_[1] ], $_[2], $_[0] ];
}

sub finalize {
    my ($self) = @_;
    $self->encode($self->{logs});
}

1;
__END__

=encoding utf-8

=head1 NAME

Web::ChromeLogger - ChromeLogger for Perl

=head1 SYNOPSIS

    use Web::ChromeLogger;

    get '/', sub {
        my $logger = Web::ChromeLogger->new();
        $logger->info('hey!');
        $logger->infof('Hello, %s!', 'John');

        my $html = render_html();

        return [
            200,
            ['X-ChromeLogger-Data' => $logger->finalize()],
            $html,
        ];
    };

=head1 DESCRIPTION

Web::ChromeLogger is a ChromeLogger library for Perl5.

Chrome Logger is a Google Chrome extension for debugging server side applications in the Chrome console.

=head1 HOW IT WORKS

This module generates the string for ChromeLogger. You can send the string in 'X-ChromeLogger-Data' header in HTTP response.

For more details, please see L<ChromeLogger's Technical Specification|http://craig.is/writing/chrome-logger/techspecs>

=head1 LARGE RESPONSE HEADER

=head2 NGINX PROXY

If you are using nginx as reverse proxy, you may need to set following parameters in your configuration file:

    proxy_buffer_size   128k;
    proxy_buffers   4 256k;
    proxy_busy_buffers_size   256k;

=head2 Maximum Data Size

Chrome has a limit of C<250kb> across all headers for a single request so that is the maximum amount of encoded data you can send.

=head1 METHODS

=over 4

=item C<< my $logger = Web::ChromeLogger->new(%args) >>

Create new instance with following parameters:

=over 4

=item json_encoder (Default: C<< JSON::XS->new()->ascii(1)->convert_blessed >> )

JSON encoder object. You can use JSON::XS or JSON::PP for this.

I guess you don't need to set this parameter.

=back

=item C<< $logger->group($title: Str) >>

=item C<< $logger->groupf($format: Str, @list) >>

Push C<group>.

=item C<< $logger->group_end($title: Str) >>

=item C<< $logger->group_endf($format: Str, @list) >>

Push C<groupEnd>.

=item C<< $logger->group_collapsed($title: Str) >>

=item C<< $logger->group_collapsedf($format: Str, @list) >>

Push C<groupCollapsed>.

=item C<< $logger->info($title: Str) >>

=item C<< $logger->infof($format: Str, @list) >>

Push C<info>.

=item C<< $logger->warn($title: Str) >>

=item C<< $logger->warnf($format: Str, @list) >>

Push C<warn>.

=item C<< $logger->error($title: Str) >>

=item C<< $logger->errorf($format: Str, @list) >>

Push C<error>.

=item C<< $logger->wrap_by_group($title: Str) >>

=item C<< $logger->wrap_by_groupf($format: Str, @list) >>

Wrap current logging data by C<$title> group.

=item C<< $logger->finalize() >>

Generate header string.

=back

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom@gmail.comE<gt>

moznion

=head1 SEE ALSO

L<ChromeLogger|http://craig.is/writing/chrome-logger>

=cut

